lib/active_model/serializer/adapter/json_sideload.rb

```diff
@@ -0,0 +1,185 @@
+module ActiveModel
+  class Serializer
+    class Adapter
+      class JsonSideload < Json
+
+        # the list of what has already been serialized will be kept here to
+        # help avoid infinite recursion
+        #
+        # this should be a list of association_name to a list of objects
+        attr_accessor :associations
+
+        # When we are sideloading associations, we can more easily track what
+        # has been serialized, so that we avoid infinite
+        # recursion / serialization.
+        def initialize(*args)
+          @associations = {}
+
+          super(*args)
+        end
+
+        def serializable_hash(options = nil)
+          options ||= {}
+
+          serialize_hash(options)
+
+          singularize_lone_associations
+
+          @associations
+        end
+
+
+        def serialize_hash(options)
+          result = nil
+
+          if serializer.respond_to?(:each)
+            #TODO: Is this ever hit?
+            # binding.pry
+            ap "wwwwwwwwwwwwwwwwwwwwww"
+            result = serialize_array(serializer, options)
```

Comment:
I think it would be very beneficial to be able to have customizable collection adapters and serializers.

I know back when I was doing a versioned API, the collections were formatted different, and that would have been pretty helpful. :-)